{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "initial_id",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:08:02.907931Z",
     "start_time": "2025-10-17T21:08:02.906106Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "762b32fa9b5cf3eb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:08:03.382037Z",
     "start_time": "2025-10-17T21:08:03.343260Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>type</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8TF40</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q1ENB6</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P36001</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MDDISGRQTLPRINRLLEHVGNPQDSLSILHIAGTNGKETVSKFLT...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P55317</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLGTVKMEGHETSDWNSYYADTQEAYSSVPVSNMNSGLGSMNSMNT...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P35583</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLGAVKMEGHEPSDWSSYYAEPEGYSSVSNMNAGLGMNGMNTYMSM...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25688</th>\n",
       "      <td>P9WQP3</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25689</th>\n",
       "      <td>P9WQP2</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25690</th>\n",
       "      <td>Q5HV12</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25691</th>\n",
       "      <td>A8FLJ3</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25692</th>\n",
       "      <td>B3GZ85</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MNRFTRYDVTPEAIFNQRRQIIKAMGLGAAALSLPNIGFAAEKSDQ...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25693 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id   kingdom   type  \\\n",
       "0         Q8TF40   EUKARYA  NO_SP   \n",
       "1         Q1ENB6   EUKARYA  NO_SP   \n",
       "2         P36001   EUKARYA  NO_SP   \n",
       "3         P55317   EUKARYA  NO_SP   \n",
       "4         P35583   EUKARYA  NO_SP   \n",
       "...          ...       ...    ...   \n",
       "25688     P9WQP3  POSITIVE    TAT   \n",
       "25689     P9WQP2  POSITIVE    TAT   \n",
       "25690     Q5HV12  NEGATIVE    TAT   \n",
       "25691     A8FLJ3  NEGATIVE    TAT   \n",
       "25692     B3GZ85  NEGATIVE    TAT   \n",
       "\n",
       "                                                sequence  \\\n",
       "0      MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...   \n",
       "1      MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...   \n",
       "2      MDDISGRQTLPRINRLLEHVGNPQDSLSILHIAGTNGKETVSKFLT...   \n",
       "3      MLGTVKMEGHETSDWNSYYADTQEAYSSVPVSNMNSGLGSMNSMNT...   \n",
       "4      MLGAVKMEGHEPSDWSSYYAEPEGYSSVSNMNAGLGMNGMNTYMSM...   \n",
       "...                                                  ...   \n",
       "25688  MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...   \n",
       "25689  MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...   \n",
       "25690  MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...   \n",
       "25691  MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...   \n",
       "25692  MNRFTRYDVTPEAIFNQRRQIIKAMGLGAAALSLPNIGFAAEKSDQ...   \n",
       "\n",
       "                                                  labels  length  \n",
       "0      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "1      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "2      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "3      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "4      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "...                                                  ...     ...  \n",
       "25688  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...      70  \n",
       "25689  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...      70  \n",
       "25690  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...      70  \n",
       "25691  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...      70  \n",
       "25692  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOO...      70  \n",
       "\n",
       "[25693 rows x 6 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p6_dataframe = pd.read_csv(\"raw/complete_set.csv\").drop(columns=[\"partition\"])\n",
    "p6_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "e4f9165d287be8e3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:08:04.033194Z",
     "start_time": "2025-10-17T21:08:04.028532Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>type</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8TF40</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q1ENB6</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P36001</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MDDISGRQTLPRINRLLEHVGNPQDSLSILHIAGTNGKETVSKFLT...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P55317</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLGTVKMEGHETSDWNSYYADTQEAYSSVPVSNMNSGLGSMNSMNT...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P35583</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLGAVKMEGHEPSDWSSYYAEPEGYSSVSNMNAGLGMNGMNTYMSM...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25688</th>\n",
       "      <td>P9WQP3</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25689</th>\n",
       "      <td>P9WQP2</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25690</th>\n",
       "      <td>Q5HV12</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25691</th>\n",
       "      <td>A8FLJ3</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25692</th>\n",
       "      <td>B3GZ85</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MNRFTRYDVTPEAIFNQRRQIIKAMGLGAAALSLPNIGFAAEKSDQ...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25449 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id   kingdom   type  \\\n",
       "0         Q8TF40   EUKARYA  NO_SP   \n",
       "1         Q1ENB6   EUKARYA  NO_SP   \n",
       "2         P36001   EUKARYA  NO_SP   \n",
       "3         P55317   EUKARYA  NO_SP   \n",
       "4         P35583   EUKARYA  NO_SP   \n",
       "...          ...       ...    ...   \n",
       "25688     P9WQP3  POSITIVE    TAT   \n",
       "25689     P9WQP2  POSITIVE    TAT   \n",
       "25690     Q5HV12  NEGATIVE    TAT   \n",
       "25691     A8FLJ3  NEGATIVE    TAT   \n",
       "25692     B3GZ85  NEGATIVE    TAT   \n",
       "\n",
       "                                                sequence  \\\n",
       "0      MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...   \n",
       "1      MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...   \n",
       "2      MDDISGRQTLPRINRLLEHVGNPQDSLSILHIAGTNGKETVSKFLT...   \n",
       "3      MLGTVKMEGHETSDWNSYYADTQEAYSSVPVSNMNSGLGSMNSMNT...   \n",
       "4      MLGAVKMEGHEPSDWSSYYAEPEGYSSVSNMNAGLGMNGMNTYMSM...   \n",
       "...                                                  ...   \n",
       "25688  MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...   \n",
       "25689  MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...   \n",
       "25690  MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...   \n",
       "25691  MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...   \n",
       "25692  MNRFTRYDVTPEAIFNQRRQIIKAMGLGAAALSLPNIGFAAEKSDQ...   \n",
       "\n",
       "                                                  labels  length  \n",
       "0      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "1      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "2      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "3      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "4      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "...                                                  ...     ...  \n",
       "25688  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...      70  \n",
       "25689  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...      70  \n",
       "25690  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...      70  \n",
       "25691  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...      70  \n",
       "25692  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOO...      70  \n",
       "\n",
       "[25449 rows x 6 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Filter Data\n",
    "# filter everything len(label) < 70\n",
    "filtered_df = p6_dataframe[p6_dataframe[\"length\"] == 70]\n",
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8f6233de9a449b42",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:08:04.701076Z",
     "start_time": "2025-10-17T21:08:04.689862Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>type</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8TF40</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q1ENB6</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P36001</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MDDISGRQTLPRINRLLEHVGNPQDSLSILHIAGTNGKETVSKFLT...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P55317</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLGTVKMEGHETSDWNSYYADTQEAYSSVPVSNMNSGLGSMNSMNT...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P35583</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLGAVKMEGHEPSDWSSYYAEPEGYSSVSNMNAGLGMNGMNTYMSM...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25444</th>\n",
       "      <td>P9WQP3</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25445</th>\n",
       "      <td>P9WQP2</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25446</th>\n",
       "      <td>Q5HV12</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25447</th>\n",
       "      <td>A8FLJ3</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25448</th>\n",
       "      <td>B3GZ85</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MNRFTRYDVTPEAIFNQRRQIIKAMGLGAAALSLPNIGFAAEKSDQ...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25449 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id   kingdom   type  \\\n",
       "0         Q8TF40   EUKARYA  NO_SP   \n",
       "1         Q1ENB6   EUKARYA  NO_SP   \n",
       "2         P36001   EUKARYA  NO_SP   \n",
       "3         P55317   EUKARYA  NO_SP   \n",
       "4         P35583   EUKARYA  NO_SP   \n",
       "...          ...       ...    ...   \n",
       "25444     P9WQP3  POSITIVE    TAT   \n",
       "25445     P9WQP2  POSITIVE    TAT   \n",
       "25446     Q5HV12  NEGATIVE    TAT   \n",
       "25447     A8FLJ3  NEGATIVE    TAT   \n",
       "25448     B3GZ85  NEGATIVE    TAT   \n",
       "\n",
       "                                                sequence  \\\n",
       "0      MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...   \n",
       "1      MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...   \n",
       "2      MDDISGRQTLPRINRLLEHVGNPQDSLSILHIAGTNGKETVSKFLT...   \n",
       "3      MLGTVKMEGHETSDWNSYYADTQEAYSSVPVSNMNSGLGSMNSMNT...   \n",
       "4      MLGAVKMEGHEPSDWSSYYAEPEGYSSVSNMNAGLGMNGMNTYMSM...   \n",
       "...                                                  ...   \n",
       "25444  MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...   \n",
       "25445  MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSR...   \n",
       "25446  MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...   \n",
       "25447  MNRRDFIKNTAIASAASVAGLSVPSSMLGAQEEDWKWDKAVCRFCG...   \n",
       "25448  MNRFTRYDVTPEAIFNQRRQIIKAMGLGAAALSLPNIGFAAEKSDQ...   \n",
       "\n",
       "                                                  labels  length  \n",
       "0      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "1      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "2      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "3      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "4      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "...                                                  ...     ...  \n",
       "25444  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...      70  \n",
       "25445  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...      70  \n",
       "25446  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...      70  \n",
       "25447  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOO...      70  \n",
       "25448  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOO...      70  \n",
       "\n",
       "[25449 rows x 6 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove duplicates\n",
    "filtered_df = filtered_df.drop_duplicates(keep=\"first\").reset_index(drop=True)\n",
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "1b0d75badb037a25",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:08:06.470452Z",
     "start_time": "2025-10-17T21:08:06.464409Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>type</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8TF40</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q1ENB6</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P36001</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MDDISGRQTLPRINRLLEHVGNPQDSLSILHIAGTNGKETVSKFLT...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P55317</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLGTVKMEGHETSDWNSYYADTQEAYSSVPVSNMNSGLGSMNSMNT...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P35583</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLGAVKMEGHEPSDWSSYYAEPEGYSSVSNMNAGLGMNGMNTYMSM...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23568</th>\n",
       "      <td>G8QMC2</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MLIRRPPDLLPSEITPEPLARGRRALLKGLGAGAALAGLGLPQISQ...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23569</th>\n",
       "      <td>H8ZPX1</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MANDKGDISKDGVSRRKFLGGAVIGAAAAAGVGSQILSLSATAQGA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23570</th>\n",
       "      <td>Q9AD93</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MRQATRKGLMTMAAATGVIAAAGGAAHADSGAHGTSSGSPGVLSGN...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23571</th>\n",
       "      <td>P46911</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MGGKHDISRRQFLNYTLTGVGGFMAASMLMPMVRFALDPVLKSTGK...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23572</th>\n",
       "      <td>Q45657</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MSDNKHRVTRRQFLNYTLTGVGGFMAAGMLMPMLRFAFDPILRETA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23573 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id   kingdom   type  \\\n",
       "0         Q8TF40   EUKARYA  NO_SP   \n",
       "1         Q1ENB6   EUKARYA  NO_SP   \n",
       "2         P36001   EUKARYA  NO_SP   \n",
       "3         P55317   EUKARYA  NO_SP   \n",
       "4         P35583   EUKARYA  NO_SP   \n",
       "...          ...       ...    ...   \n",
       "23568     G8QMC2  NEGATIVE    TAT   \n",
       "23569     H8ZPX1  NEGATIVE    TAT   \n",
       "23570     Q9AD93  POSITIVE    TAT   \n",
       "23571     P46911  POSITIVE    TAT   \n",
       "23572     Q45657  POSITIVE    TAT   \n",
       "\n",
       "                                                sequence  \\\n",
       "0      MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...   \n",
       "1      MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...   \n",
       "2      MDDISGRQTLPRINRLLEHVGNPQDSLSILHIAGTNGKETVSKFLT...   \n",
       "3      MLGTVKMEGHETSDWNSYYADTQEAYSSVPVSNMNSGLGSMNSMNT...   \n",
       "4      MLGAVKMEGHEPSDWSSYYAEPEGYSSVSNMNAGLGMNGMNTYMSM...   \n",
       "...                                                  ...   \n",
       "23568  MLIRRPPDLLPSEITPEPLARGRRALLKGLGAGAALAGLGLPQISQ...   \n",
       "23569  MANDKGDISKDGVSRRKFLGGAVIGAAAAAGVGSQILSLSATAQGA...   \n",
       "23570  MRQATRKGLMTMAAATGVIAAAGGAAHADSGAHGTSSGSPGVLSGN...   \n",
       "23571  MGGKHDISRRQFLNYTLTGVGGFMAASMLMPMVRFALDPVLKSTGK...   \n",
       "23572  MSDNKHRVTRRQFLNYTLTGVGGFMAAGMLMPMLRFAFDPILRETA...   \n",
       "\n",
       "                                                  labels  length  \n",
       "0      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "1      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "2      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "3      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "4      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "...                                                  ...     ...  \n",
       "23568  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT...      70  \n",
       "23569  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...      70  \n",
       "23570  TTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOOOO...      70  \n",
       "23571  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOO...      70  \n",
       "23572  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOO...      70  \n",
       "\n",
       "[23573 rows x 6 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove entries with differing annotations\n",
    "filtered_df = filtered_df.drop_duplicates(subset=[\"sequence\"], keep=False).reset_index(drop=True)\n",
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "f9e0a019c5e9c882",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-16T19:02:46.997810Z",
     "start_time": "2025-10-16T19:02:46.763074Z"
    }
   },
   "outputs": [],
   "source": [
    "with open(\"input.fasta\", \"w\") as f:\n",
    "    for i, row in filtered_df.iterrows():\n",
    "        f.write(f\">{row['uniprot_id']}\\n{row['sequence']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "17720be40f0a6901",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-16T19:02:47.003164Z",
     "start_time": "2025-10-16T19:02:47.001194Z"
    },
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb input.fasta inputDB \n",
      "\n",
      "MMseqs Version:                    \tbd01c2229f027d8d8e61947f44d11ef1a7669212\n",
      "Database type                      \t0\n",
      "Shuffle input database             \ttrue\n",
      "Createdb mode                      \t0\n",
      "Write lookup file                  \t1\n",
      "Offset of numeric ids              \t0\n",
      "Threads                            \t12\n",
      "Compressed                         \t0\n",
      "Mask residues                      \t0\n",
      "Mask residues probability          \t0.9\n",
      "Mask lower case residues           \t0\n",
      "Mask lower letter repeating N times\t0\n",
      "Use GPU                            \t0\n",
      "Verbosity                          \t3\n",
      "\n",
      "Converting sequences\n",
      "[==\n",
      "Time for merging to inputDB_h: 0h 0m 0s 1ms\n",
      "Time for merging to inputDB: 0h 0m 0s 1ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 24ms\n",
      "Create directory tmp\n",
      "cluster inputDB clusterRes tmp --min-seq-id 0.3 \n",
      "\n",
      "MMseqs Version:                     \tbd01c2229f027d8d8e61947f44d11ef1a7669212\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias scale            \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Mask lower letter repeating N times \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t12\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Set mode                            \tfalse\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \tfalse\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 3\n",
      "linclust inputDB tmp/15442017313262073528/clu_redundancy tmp/15442017313262073528/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 12 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --set-mode 0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 0 --force-reuse 0 \n",
      "\n",
      "kmermatcher inputDB tmp/15442017313262073528/linclust/7459307913141530178/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 12 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher inputDB tmp/15442017313262073528/linclust/7459307913141530178/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 12 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 23573 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 23.57K 0s 5ms\n",
      "Sort kmer 0h 0m 0s 29ms\n",
      "Sort by rep. sequence 0h 0m 0s 1ms\n",
      "Time for fill: 0h 0m 0s 0ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 47ms\n",
      "rescorediagonal inputDB inputDB tmp/15442017313262073528/linclust/7459307913141530178/pref tmp/15442017313262073528/linclust/7459307913141530178/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 23.57K 0s 2ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 24ms\n",
      "clust inputDB tmp/15442017313262073528/linclust/7459307913141530178/pref_rescore1 tmp/15442017313262073528/linclust/7459307913141530178/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 12 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --set-mode 0 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 23.57K 0s 0ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 5457 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 23.57K 0s 0ms\n",
      "Add missing connections\n",
      "[=================================================================] 23.57K 0s 0ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 1ms\n",
      "Total time: 0h 0m 0s 25ms\n",
      "\n",
      "Size of the sequence database: 23573\n",
      "Size of the alignment database: 23573\n",
      "Number of clusters: 19769\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 31ms\n",
      "createsubdb tmp/15442017313262073528/linclust/7459307913141530178/order_redundancy inputDB tmp/15442017313262073528/linclust/7459307913141530178/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "createsubdb tmp/15442017313262073528/linclust/7459307913141530178/order_redundancy tmp/15442017313262073528/linclust/7459307913141530178/pref tmp/15442017313262073528/linclust/7459307913141530178/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "filterdb tmp/15442017313262073528/linclust/7459307913141530178/pref_filter1 tmp/15442017313262073528/linclust/7459307913141530178/pref_filter2 --filter-file tmp/15442017313262073528/linclust/7459307913141530178/order_redundancy --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 19.77K 0s 3ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "rescorediagonal tmp/15442017313262073528/linclust/7459307913141530178/input_step_redundancy tmp/15442017313262073528/linclust/7459307913141530178/input_step_redundancy tmp/15442017313262073528/linclust/7459307913141530178/pref_filter2 tmp/15442017313262073528/linclust/7459307913141530178/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 19.77K 0s 3ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "align tmp/15442017313262073528/linclust/7459307913141530178/input_step_redundancy tmp/15442017313262073528/linclust/7459307913141530178/input_step_redundancy tmp/15442017313262073528/linclust/7459307913141530178/pref_rescore2 tmp/15442017313262073528/linclust/7459307913141530178/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 19769 type: Aminoacid\n",
      "Target database size: 19769 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 19.77K 0s 24ms\n",
      "Time for merging to aln: 0h 0m 0s 3ms\n",
      "21680 alignments calculated\n",
      "20288 sequence pairs passed the thresholds (0.935793 of overall calculated)\n",
      "1.026253 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 31ms\n",
      "clust tmp/15442017313262073528/linclust/7459307913141530178/input_step_redundancy tmp/15442017313262073528/linclust/7459307913141530178/aln tmp/15442017313262073528/linclust/7459307913141530178/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 12 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --set-mode 0 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 19.77K 0s 3ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 519 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 19.77K 0s 3ms\n",
      "Add missing connections\n",
      "[=================================================================] 19.77K 0s 0ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 28ms\n",
      "Total time: 0h 0m 0s 33ms\n",
      "\n",
      "Size of the sequence database: 19769\n",
      "Size of the alignment database: 19769\n",
      "Number of clusters: 19312\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 36ms\n",
      "mergeclusters inputDB tmp/15442017313262073528/clu_redundancy tmp/15442017313262073528/linclust/7459307913141530178/pre_clust tmp/15442017313262073528/linclust/7459307913141530178/clust --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 19.77K 0s 3ms\n",
      "Clustering step 2\n",
      "[=================================================================] 19.31K 0s 11ms\n",
      "Write merged clustering\n",
      "[=================================================================] 23.57K 0s 12ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "createsubdb tmp/15442017313262073528/clu_redundancy inputDB tmp/15442017313262073528/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "prefilter tmp/15442017313262073528/input_step_redundancy tmp/15442017313262073528/input_step_redundancy tmp/15442017313262073528/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 19312 type: Aminoacid\n",
      "Estimated memory consumption: 998M\n",
      "Target database size: 19312 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 19.31K 0s 8ms\n",
      "Index table: Masked residues: 23620\n",
      "Index table: fill\n",
      "[=================================================================] 19.31K 0s 8ms\n",
      "Index statistics\n",
      "Entries:          520604\n",
      "DB size:          491 MB\n",
      "Avg k-mer size:   0.008134\n",
      "Top 10 k-mers\n",
      "    GFFPEE\t10\n",
      "    FRHTEL\t10\n",
      "    DADCVH\t9\n",
      "    LGGFVR\t9\n",
      "    LGGFVY\t9\n",
      "    AVACYF\t8\n",
      "    IGGFVY\t8\n",
      "    VGGYVY\t8\n",
      "    FVFDDA\t7\n",
      "    RRFKGA\t7\n",
      "Time for index table init: 0h 0m 0s 201ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 19312\n",
      "Target db start 1 to 19312\n",
      "[=================================================================] 19.31K 0s 21ms\n",
      "\n",
      "1.714636 k-mers per position\n",
      "29 DB matches per sequence\n",
      "0 overflows\n",
      "1 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 396ms\n",
      "align tmp/15442017313262073528/input_step_redundancy tmp/15442017313262073528/input_step_redundancy tmp/15442017313262073528/pref_step0 tmp/15442017313262073528/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 19312 type: Aminoacid\n",
      "Target database size: 19312 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 19.31K 0s 20ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 2ms\n",
      "24227 alignments calculated\n",
      "21550 sequence pairs passed the thresholds (0.889503 of overall calculated)\n",
      "1.115886 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 26ms\n",
      "clust tmp/15442017313262073528/input_step_redundancy tmp/15442017313262073528/aln_step0 tmp/15442017313262073528/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 12 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --set-mode 0 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 19.31K 0s 0ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 8 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 19.31K 0s 0ms\n",
      "Add missing connections\n",
      "[=================================================================] 19.31K 0s 0ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 1ms\n",
      "Total time: 0h 0m 0s 3ms\n",
      "\n",
      "Size of the sequence database: 19312\n",
      "Size of the alignment database: 19312\n",
      "Number of clusters: 18383\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 13ms\n",
      "createsubdb tmp/15442017313262073528/clu_step0 tmp/15442017313262073528/input_step_redundancy tmp/15442017313262073528/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "prefilter tmp/15442017313262073528/input_step1 tmp/15442017313262073528/input_step1 tmp/15442017313262073528/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 18383 type: Aminoacid\n",
      "Estimated memory consumption: 997M\n",
      "Target database size: 18383 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 18.38K 0s 8ms\n",
      "Index table: Masked residues: 22973\n",
      "Index table: fill\n",
      "[=================================================================] 18.38K 0s 18ms\n",
      "Index statistics\n",
      "Entries:          1012445\n",
      "DB size:          494 MB\n",
      "Avg k-mer size:   0.015819\n",
      "Top 10 k-mers\n",
      "    LGGFVK\t9\n",
      "    MKLLLV\t9\n",
      "    LGGFVY\t9\n",
      "    GAGGAA\t8\n",
      "    LGGFVR\t8\n",
      "    IGGFVY\t8\n",
      "    VGGYVY\t8\n",
      "    GFLELA\t7\n",
      "    TGGSSG\t7\n",
      "    VVGGGL\t7\n",
      "Time for index table init: 0h 0m 0s 212ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 18383\n",
      "Target db start 1 to 18383\n",
      "[=================================================================] 18.38K 0s 128ms\n",
      "\n",
      "15.509256 k-mers per position\n",
      "58 DB matches per sequence\n",
      "0 overflows\n",
      "1 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 515ms\n",
      "align tmp/15442017313262073528/input_step1 tmp/15442017313262073528/input_step1 tmp/15442017313262073528/pref_step1 tmp/15442017313262073528/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 18383 type: Aminoacid\n",
      "Target database size: 18383 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 18.38K 0s 27ms\n",
      "Time for merging to aln_step1: 0h 0m 0s 2ms\n",
      "29585 alignments calculated\n",
      "19829 sequence pairs passed the thresholds (0.670238 of overall calculated)\n",
      "1.078660 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 33ms\n",
      "clust tmp/15442017313262073528/input_step1 tmp/15442017313262073528/aln_step1 tmp/15442017313262073528/clu_step1 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 12 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --set-mode 0 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 18.38K 0s 3ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 270 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 18.38K 0s 3ms\n",
      "Add missing connections\n",
      "[=================================================================] 18.38K 0s 0ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 14ms\n",
      "Total time: 0h 0m 0s 25ms\n",
      "\n",
      "Size of the sequence database: 18383\n",
      "Size of the alignment database: 18383\n",
      "Number of clusters: 17626\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 36ms\n",
      "createsubdb tmp/15442017313262073528/clu_step1 tmp/15442017313262073528/input_step1 tmp/15442017313262073528/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "prefilter tmp/15442017313262073528/input_step2 tmp/15442017313262073528/input_step2 tmp/15442017313262073528/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 5 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 17626 type: Aminoacid\n",
      "Estimated memory consumption: 997M\n",
      "Target database size: 17626 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 17.63K 0s 7ms\n",
      "Index table: Masked residues: 22662\n",
      "Index table: fill\n",
      "[=================================================================] 17.63K 0s 13ms\n",
      "Index statistics\n",
      "Entries:          1037919\n",
      "DB size:          494 MB\n",
      "Avg k-mer size:   0.016217\n",
      "Top 10 k-mers\n",
      "    LLLLLL\t20\n",
      "    LLLLAA\t18\n",
      "    LLLLLA\t14\n",
      "    LAAALA\t9\n",
      "    AALAAL\t9\n",
      "    LLLALL\t9\n",
      "    ALLLLL\t9\n",
      "    LLVLLL\t9\n",
      "    LLLLSL\t9\n",
      "    LGGFVY\t9\n",
      "Time for index table init: 0h 0m 0s 203ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 17626\n",
      "Target db start 1 to 17626\n",
      "[=================================================================] 17.63K 0s 526ms\n",
      "\n",
      "117.586937 k-mers per position\n",
      "167 DB matches per sequence\n",
      "0 overflows\n",
      "4 sequences passed prefiltering per query sequence\n",
      "4 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 890ms\n",
      "align tmp/15442017313262073528/input_step2 tmp/15442017313262073528/input_step2 tmp/15442017313262073528/pref_step2 tmp/15442017313262073528/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 17626 type: Aminoacid\n",
      "Target database size: 17626 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 17.63K 0s 35ms\n",
      "Time for merging to aln_step2: 0h 0m 0s 2ms\n",
      "75678 alignments calculated\n",
      "18545 sequence pairs passed the thresholds (0.245051 of overall calculated)\n",
      "1.052139 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 40ms\n",
      "clust tmp/15442017313262073528/input_step2 tmp/15442017313262073528/aln_step2 tmp/15442017313262073528/clu_step2 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 12 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --set-mode 0 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 17.63K 0s 2ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 223 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 17.63K 0s 5ms\n",
      "Add missing connections\n",
      "[=================================================================] 17.63K 0s 0ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 27ms\n",
      "Total time: 0h 0m 0s 59ms\n",
      "\n",
      "Size of the sequence database: 17626\n",
      "Size of the alignment database: 17626\n",
      "Number of clusters: 17094\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 62ms\n",
      "mergeclusters inputDB clusterRes tmp/15442017313262073528/clu_redundancy tmp/15442017313262073528/clu_step0 tmp/15442017313262073528/clu_step1 tmp/15442017313262073528/clu_step2 --threads 12 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 19.31K 0s 0ms\n",
      "Clustering step 2\n",
      "[=================================================================] 18.38K 0s 5ms\n",
      "Clustering step 3\n",
      "[=================================================================] 17.63K 0s 12ms\n",
      "Clustering step 4\n",
      "[=================================================================] 17.09K 0s 15ms\n",
      "Write merged clustering\n",
      "[=================================================================] 23.57K 0s 28ms\n",
      "Time for merging to clusterRes: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 36ms\n",
      "createsubdb clusterRes inputDB repDB \n",
      "\n",
      "MMseqs Version: \tbd01c2229f027d8d8e61947f44d11ef1a7669212\n",
      "Subdb mode      \t0\n",
      "Database ID mode\t0\n",
      "Verbosity       \t3\n",
      "\n",
      "Time for merging to repDB: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "convert2fasta repDB reduced_30.fasta \n",
      "\n",
      "MMseqs Version:\tbd01c2229f027d8d8e61947f44d11ef1a7669212\n",
      "Use header DB\tfalse\n",
      "Verbosity    \t3\n",
      "\n",
      "Start writing file to reduced_30.fasta\n",
      "Time for processing: 0h 0m 0s 2ms\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "## external clustering using mmseqs\n",
    "mmseqs createdb input.fasta inputDB\n",
    "mmseqs cluster inputDB clusterRes tmp --min-seq-id 0.3\n",
    "mmseqs createsubdb clusterRes inputDB repDB\n",
    "mmseqs convert2fasta repDB reduced_30.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "7fced33ca604a774",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:23:32.837639Z",
     "start_time": "2025-10-17T21:23:32.822055Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>type</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8TF40</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q1ENB6</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q9PSY4</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MTLSSDMSDVLAEETDIDVVGEEDEPRAEEEEEEDGELLMPRSPHC...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P55316</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLDMGDRKEVKMIPKSSFSINSLVPEAVQNDNHHASHGHHNSHHPQ...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q9BZS1</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MPNPRPGKPSAPSLALGPSPGASPSWRAAPKASDLLGARGPGGTFQ...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23565</th>\n",
       "      <td>C8WLM1</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TATLIPO</td>\n",
       "      <td>MEYGKCRGIERGMGRRDFLKAATLLGATAAGAGMLAGCAPKSASEA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23566</th>\n",
       "      <td>Q8VPD4</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MAFTRRSFMKGLGATGGAGLAYGAMSTLGLAPSTAAPARTFQPLAA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23567</th>\n",
       "      <td>Q8A861</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MPNRRDFLKTAAFATLGSGIAVSQVLAGECMPSAIHINKYGIGGKM...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23569</th>\n",
       "      <td>H8ZPX1</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MANDKGDISKDGVSRRKFLGGAVIGAAAAAGVGSQILSLSATAQGA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23572</th>\n",
       "      <td>Q45657</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MSDNKHRVTRRQFLNYTLTGVGGFMAAGMLMPMLRFAFDPILRETA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17094 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id   kingdom     type  \\\n",
       "0         Q8TF40   EUKARYA    NO_SP   \n",
       "1         Q1ENB6   EUKARYA    NO_SP   \n",
       "5         Q9PSY4   EUKARYA    NO_SP   \n",
       "6         P55316   EUKARYA    NO_SP   \n",
       "7         Q9BZS1   EUKARYA    NO_SP   \n",
       "...          ...       ...      ...   \n",
       "23565     C8WLM1  POSITIVE  TATLIPO   \n",
       "23566     Q8VPD4  POSITIVE      TAT   \n",
       "23567     Q8A861  NEGATIVE      TAT   \n",
       "23569     H8ZPX1  NEGATIVE      TAT   \n",
       "23572     Q45657  POSITIVE      TAT   \n",
       "\n",
       "                                                sequence  \\\n",
       "0      MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...   \n",
       "1      MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...   \n",
       "5      MTLSSDMSDVLAEETDIDVVGEEDEPRAEEEEEEDGELLMPRSPHC...   \n",
       "6      MLDMGDRKEVKMIPKSSFSINSLVPEAVQNDNHHASHGHHNSHHPQ...   \n",
       "7      MPNPRPGKPSAPSLALGPSPGASPSWRAAPKASDLLGARGPGGTFQ...   \n",
       "...                                                  ...   \n",
       "23565  MEYGKCRGIERGMGRRDFLKAATLLGATAAGAGMLAGCAPKSASEA...   \n",
       "23566  MAFTRRSFMKGLGATGGAGLAYGAMSTLGLAPSTAAPARTFQPLAA...   \n",
       "23567  MPNRRDFLKTAAFATLGSGIAVSQVLAGECMPSAIHINKYGIGGKM...   \n",
       "23569  MANDKGDISKDGVSRRKFLGGAVIGAAAAAGVGSQILSLSATAQGA...   \n",
       "23572  MSDNKHRVTRRQFLNYTLTGVGGFMAAGMLMPMLRFAFDPILRETA...   \n",
       "\n",
       "                                                  labels  length  \n",
       "0      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "1      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "5      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "6      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "7      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70  \n",
       "...                                                  ...     ...  \n",
       "23565  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOO...      70  \n",
       "23566  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTO...      70  \n",
       "23567  TTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOOOOO...      70  \n",
       "23569  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...      70  \n",
       "23572  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOO...      70  \n",
       "\n",
       "[17094 rows x 6 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "representatives = set()\n",
    "with open(\"reduced_30.fasta\", \"r\") as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\">\"):\n",
    "            representatives.add(line[1:].strip())\n",
    "\n",
    "reduced_df = filtered_df[filtered_df[\"uniprot_id\"].isin(representatives)].copy()\n",
    "reduced_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "84d47fed095a0821",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:23:34.492052Z",
     "start_time": "2025-10-17T21:23:34.299680Z"
    }
   },
   "outputs": [],
   "source": [
    "## amino acid validation\n",
    "valid_aa = set(\"ACDEFGHIKLMNPQRSTVWY\")\n",
    "\n",
    "for i, row in reduced_df.iterrows():\n",
    "    if set(row[\"sequence\"]) - valid_aa:\n",
    "        print(f\"{i}; {row['uniprot_id']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "2a035a5a45f50062",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:23:34.859857Z",
     "start_time": "2025-10-17T21:23:34.855260Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>type</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "      <th>length</th>\n",
       "      <th>has_sp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8TF40</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q1ENB6</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q9PSY4</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MTLSSDMSDVLAEETDIDVVGEEDEPRAEEEEEEDGELLMPRSPHC...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P55316</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLDMGDRKEVKMIPKSSFSINSLVPEAVQNDNHHASHGHHNSHHPQ...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Q9BZS1</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MPNPRPGKPSAPSLALGPSPGASPSWRAAPKASDLLGARGPGGTFQ...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23565</th>\n",
       "      <td>C8WLM1</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TATLIPO</td>\n",
       "      <td>MEYGKCRGIERGMGRRDFLKAATLLGATAAGAGMLAGCAPKSASEA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23566</th>\n",
       "      <td>Q8VPD4</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MAFTRRSFMKGLGATGGAGLAYGAMSTLGLAPSTAAPARTFQPLAA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTO...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23567</th>\n",
       "      <td>Q8A861</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MPNRRDFLKTAAFATLGSGIAVSQVLAGECMPSAIHINKYGIGGKM...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23569</th>\n",
       "      <td>H8ZPX1</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MANDKGDISKDGVSRRKFLGGAVIGAAAAAGVGSQILSLSATAQGA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23572</th>\n",
       "      <td>Q45657</td>\n",
       "      <td>POSITIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MSDNKHRVTRRQFLNYTLTGVGGFMAAGMLMPMLRFAFDPILRETA...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17094 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id   kingdom     type  \\\n",
       "0         Q8TF40   EUKARYA    NO_SP   \n",
       "1         Q1ENB6   EUKARYA    NO_SP   \n",
       "5         Q9PSY4   EUKARYA    NO_SP   \n",
       "6         P55316   EUKARYA    NO_SP   \n",
       "7         Q9BZS1   EUKARYA    NO_SP   \n",
       "...          ...       ...      ...   \n",
       "23565     C8WLM1  POSITIVE  TATLIPO   \n",
       "23566     Q8VPD4  POSITIVE      TAT   \n",
       "23567     Q8A861  NEGATIVE      TAT   \n",
       "23569     H8ZPX1  NEGATIVE      TAT   \n",
       "23572     Q45657  POSITIVE      TAT   \n",
       "\n",
       "                                                sequence  \\\n",
       "0      MAPTLFQKLFSKRTGLGAPGRDARDPDCGFSWPLPEFDPSQIRLIV...   \n",
       "1      MDFTSLETTTFEEVVIALGSNVGNRMNNFKEALRLMKDYGISVTRH...   \n",
       "5      MTLSSDMSDVLAEETDIDVVGEEDEPRAEEEEEEDGELLMPRSPHC...   \n",
       "6      MLDMGDRKEVKMIPKSSFSINSLVPEAVQNDNHHASHGHHNSHHPQ...   \n",
       "7      MPNPRPGKPSAPSLALGPSPGASPSWRAAPKASDLLGARGPGGTFQ...   \n",
       "...                                                  ...   \n",
       "23565  MEYGKCRGIERGMGRRDFLKAATLLGATAAGAGMLAGCAPKSASEA...   \n",
       "23566  MAFTRRSFMKGLGATGGAGLAYGAMSTLGLAPSTAAPARTFQPLAA...   \n",
       "23567  MPNRRDFLKTAAFATLGSGIAVSQVLAGECMPSAIHINKYGIGGKM...   \n",
       "23569  MANDKGDISKDGVSRRKFLGGAVIGAAAAAGVGSQILSLSATAQGA...   \n",
       "23572  MSDNKHRVTRRQFLNYTLTGVGGFMAAGMLMPMLRFAFDPILRETA...   \n",
       "\n",
       "                                                  labels  length  has_sp  \n",
       "0      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "1      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "5      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "6      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "7      IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "...                                                  ...     ...     ...  \n",
       "23565  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOO...      70    True  \n",
       "23566  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTO...      70    True  \n",
       "23567  TTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOOOOOOOOOOOO...      70    True  \n",
       "23569  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOO...      70    True  \n",
       "23572  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOOOOOOOOO...      70    True  \n",
       "\n",
       "[17094 rows x 7 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## binary split for sp / no_sp classification\n",
    "reduced_df.loc[:, \"has_sp\"] = reduced_df[\"type\"] != \"NO_SP\"\n",
    "reduced_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "128285bb647605ee",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:23:35.527126Z",
     "start_time": "2025-10-17T21:23:35.520520Z"
    }
   },
   "outputs": [],
   "source": [
    "# train&val 80%, test 20%\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "train_val_df, test_df = train_test_split(reduced_df, test_size=0.2, stratify=reduced_df[\"type\"], random_state=69)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "baf545e711b4f41",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:23:35.989876Z",
     "start_time": "2025-10-17T21:23:35.984387Z"
    }
   },
   "outputs": [],
   "source": [
    "train_df, val_df = train_test_split(train_val_df, test_size=0.2, stratify=train_val_df[\"type\"], random_state=69)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "59b409a3ec9ca8fa",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:23:36.314568Z",
     "start_time": "2025-10-17T21:23:36.310511Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>type</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "      <th>length</th>\n",
       "      <th>has_sp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13293</th>\n",
       "      <td>Q9UUK4</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MFSSKRYLNSFGWEEGNALKEGGLTKPILTSRKYNTHGLGAKHDIA...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12760</th>\n",
       "      <td>O74417</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MSQVIQYSEADLLYLSKSPLIKKPENLPEWILPEAMKADRERHIRQ...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11122</th>\n",
       "      <td>P11181</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MAAALVLRTWSRAAGQLICVRYFQTCGNVHVLKPKYVCFFGYPPFK...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11092</th>\n",
       "      <td>O15504</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MAICQFFLQGRCRFGDRCWNEHPGARGAGGGRQQPQQQPSGNNRRG...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5950</th>\n",
       "      <td>Q06266</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MSSQFPSSPYRTVDPYSPPNYKQQPNCPSSNYEKAGKTASESIGNF...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9875</th>\n",
       "      <td>Q8N9N8</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MSQATKRKHVVKEVLGEHIVPSDQQQIVRVLRTPGNNLHEVETAQG...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11130</th>\n",
       "      <td>Q9P4X2</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MYFLILFSSNAELGEVLRTMRSVEDRFNKRFHYPWTFYGMDEFTEY...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3876</th>\n",
       "      <td>Q9UJ41</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MVVVTGREPDSRRQDGAMSSSDAEDDFLEPATPTATQAGHALPLLP...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5163</th>\n",
       "      <td>P15619</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MEGGKGEGKRMKEEAPSKKLPPKIYGGDAGTPTKAAHDEILSSLLR...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6688</th>\n",
       "      <td>Q03771</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MSELEATIRQAKEALAENNAKKALKILKPFKSSLKKENANNVILNE...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10940 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id  kingdom   type  \\\n",
       "13293     Q9UUK4  EUKARYA  NO_SP   \n",
       "12760     O74417  EUKARYA  NO_SP   \n",
       "11122     P11181  EUKARYA  NO_SP   \n",
       "11092     O15504  EUKARYA  NO_SP   \n",
       "5950      Q06266  EUKARYA  NO_SP   \n",
       "...          ...      ...    ...   \n",
       "9875      Q8N9N8  EUKARYA  NO_SP   \n",
       "11130     Q9P4X2  EUKARYA  NO_SP   \n",
       "3876      Q9UJ41  EUKARYA  NO_SP   \n",
       "5163      P15619  EUKARYA  NO_SP   \n",
       "6688      Q03771  EUKARYA  NO_SP   \n",
       "\n",
       "                                                sequence  \\\n",
       "13293  MFSSKRYLNSFGWEEGNALKEGGLTKPILTSRKYNTHGLGAKHDIA...   \n",
       "12760  MSQVIQYSEADLLYLSKSPLIKKPENLPEWILPEAMKADRERHIRQ...   \n",
       "11122  MAAALVLRTWSRAAGQLICVRYFQTCGNVHVLKPKYVCFFGYPPFK...   \n",
       "11092  MAICQFFLQGRCRFGDRCWNEHPGARGAGGGRQQPQQQPSGNNRRG...   \n",
       "5950   MSSQFPSSPYRTVDPYSPPNYKQQPNCPSSNYEKAGKTASESIGNF...   \n",
       "...                                                  ...   \n",
       "9875   MSQATKRKHVVKEVLGEHIVPSDQQQIVRVLRTPGNNLHEVETAQG...   \n",
       "11130  MYFLILFSSNAELGEVLRTMRSVEDRFNKRFHYPWTFYGMDEFTEY...   \n",
       "3876   MVVVTGREPDSRRQDGAMSSSDAEDDFLEPATPTATQAGHALPLLP...   \n",
       "5163   MEGGKGEGKRMKEEAPSKKLPPKIYGGDAGTPTKAAHDEILSSLLR...   \n",
       "6688   MSELEATIRQAKEALAENNAKKALKILKPFKSSLKKENANNVILNE...   \n",
       "\n",
       "                                                  labels  length  has_sp  \n",
       "13293  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "12760  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "11122  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "11092  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "5950   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "...                                                  ...     ...     ...  \n",
       "9875   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "11130  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "3876   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "5163   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "6688   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "\n",
       "[10940 rows x 7 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "1ccb024b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "counts = train_df[\"type\"].value_counts()\n",
    "max_count = counts.max()\n",
    "\n",
    "dfs_to_concat = []\n",
    "\n",
    "for label, count in counts.items():\n",
    "    # Filter by the specific label\n",
    "    train_subset = train_df[train_df[\"type\"] == label]\n",
    "    \n",
    "    # Oversample to match max_coun replace=True so we can sample more rows than exist\n",
    "    oversampled_subset = train_subset.sample(n=max_count, replace=True, random_state=69)\n",
    "    \n",
    "    dfs_to_concat.append(oversampled_subset)\n",
    "\n",
    "# Concatenate at the end\n",
    "train_df = pd.concat(dfs_to_concat, ignore_index=True)\n",
    "\n",
    "# Shuffle the final dataframe so classes aren't ordered blocks\n",
    "train_df = train_df.sample(frac=1, random_state=42).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "361612a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "type\n",
       "PILIN      9082\n",
       "TAT        9082\n",
       "SP         9082\n",
       "TATLIPO    9082\n",
       "LIPO       9082\n",
       "NO_SP      9082\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check new class distribution\n",
    "train_df[\"type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "5461eae43025b4d3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:23:36.785154Z",
     "start_time": "2025-10-17T21:23:36.781076Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>type</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "      <th>length</th>\n",
       "      <th>has_sp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9773</th>\n",
       "      <td>Q9FNY0</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MSGVVRSSPGSSQPPPPPPHHPPSSPVPVTSTPVIPPIRRHLAFAS...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6202</th>\n",
       "      <td>A4Q9E8</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MLQCLTSESEEGAEEREESSTEDLEELKEFVTLAFVRENTQKRLQN...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2522</th>\n",
       "      <td>Q8RXH8</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MKAGMFYCASQASTATANGERTVTARAIDRHNPIIKDGRRSFTAPC...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12591</th>\n",
       "      <td>Q10200</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MKKNNERVNTNPSLISKSYNMKKRPVIGTRRHFARKEIESDSDDED...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4775</th>\n",
       "      <td>Q9UTD7</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MAKSQRSQIKNEDISFVSQEDDSSKKFVKDDSENSEEELEDLNSED...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8563</th>\n",
       "      <td>P0DJH7</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MEEREGGFRKETVERLLRLHFRDGRTRVNGDALLLMAELLKVFVRE...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21268</th>\n",
       "      <td>O75323</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MAARVLRARGAAWAGGLLQRAAPCSLLPRLRTWTSSSNRSREDSWL...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18983</th>\n",
       "      <td>P37665</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>LIPO</td>\n",
       "      <td>MKKRVYLIAAVVSGALAVSGCTTNPYTGEREAGKSAIGAGLGSLVG...</td>\n",
       "      <td>LLLLLLLLLLLLLLLLLLLLOOOOOOOOOOOOOOOOMMMMMMMMMM...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1402</th>\n",
       "      <td>Q62315</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MSKERPKRNIIQKKYDDSDGIPWSEERVVRKVLYLSLKEFKNAQKR...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9026</th>\n",
       "      <td>Q7XA78</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MNTKLSLSQTKIFTFTTWFNDTRSGLDRRSSISPTLCSKPVYSGKL...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2735 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id   kingdom   type  \\\n",
       "9773      Q9FNY0   EUKARYA  NO_SP   \n",
       "6202      A4Q9E8   EUKARYA  NO_SP   \n",
       "2522      Q8RXH8   EUKARYA  NO_SP   \n",
       "12591     Q10200   EUKARYA  NO_SP   \n",
       "4775      Q9UTD7   EUKARYA  NO_SP   \n",
       "...          ...       ...    ...   \n",
       "8563      P0DJH7   EUKARYA  NO_SP   \n",
       "21268     O75323   EUKARYA  NO_SP   \n",
       "18983     P37665  NEGATIVE   LIPO   \n",
       "1402      Q62315   EUKARYA  NO_SP   \n",
       "9026      Q7XA78   EUKARYA  NO_SP   \n",
       "\n",
       "                                                sequence  \\\n",
       "9773   MSGVVRSSPGSSQPPPPPPHHPPSSPVPVTSTPVIPPIRRHLAFAS...   \n",
       "6202   MLQCLTSESEEGAEEREESSTEDLEELKEFVTLAFVRENTQKRLQN...   \n",
       "2522   MKAGMFYCASQASTATANGERTVTARAIDRHNPIIKDGRRSFTAPC...   \n",
       "12591  MKKNNERVNTNPSLISKSYNMKKRPVIGTRRHFARKEIESDSDDED...   \n",
       "4775   MAKSQRSQIKNEDISFVSQEDDSSKKFVKDDSENSEEELEDLNSED...   \n",
       "...                                                  ...   \n",
       "8563   MEEREGGFRKETVERLLRLHFRDGRTRVNGDALLLMAELLKVFVRE...   \n",
       "21268  MAARVLRARGAAWAGGLLQRAAPCSLLPRLRTWTSSSNRSREDSWL...   \n",
       "18983  MKKRVYLIAAVVSGALAVSGCTTNPYTGEREAGKSAIGAGLGSLVG...   \n",
       "1402   MSKERPKRNIIQKKYDDSDGIPWSEERVVRKVLYLSLKEFKNAQKR...   \n",
       "9026   MNTKLSLSQTKIFTFTTWFNDTRSGLDRRSSISPTLCSKPVYSGKL...   \n",
       "\n",
       "                                                  labels  length  has_sp  \n",
       "9773   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "6202   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "2522   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "12591  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "4775   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "...                                                  ...     ...     ...  \n",
       "8563   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "21268  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "18983  LLLLLLLLLLLLLLLLLLLLOOOOOOOOOOOOOOOOMMMMMMMMMM...      70    True  \n",
       "1402   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "9026   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "\n",
       "[2735 rows x 7 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "val_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "36dd56f89f42c3ef",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:23:37.282643Z",
     "start_time": "2025-10-17T21:23:37.278713Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>type</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "      <th>length</th>\n",
       "      <th>has_sp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18084</th>\n",
       "      <td>B2VGX7</td>\n",
       "      <td>NEGATIVE</td>\n",
       "      <td>TAT</td>\n",
       "      <td>MKAVNPLTENDVTPESLFNARRRTVLKMLGMSAAALSLPGAARADL...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOO...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>Q9DEN4</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MTLSGSGSASDMSGQTVLSADDADIDVVGEGDEALDKDSECESTAG...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1049</th>\n",
       "      <td>A9QM74</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MPTLDAPEERRRKFKYRGKDVSLRRQQRMAVSLELRKAKKDEQTLK...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2541</th>\n",
       "      <td>Q8K3I4</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MGRKLDLSGLTDDETEHVLQVVQRDFNLRKKEEDRLSEMKQRLAEE...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4434</th>\n",
       "      <td>Q920M2</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MEVPELGPGLVERLEQLATCPLCGGPFEDPVLLACEHSFCRSCLAR...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4214</th>\n",
       "      <td>Q8NE09</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>NO_SP</td>\n",
       "      <td>MPEKRLTAEPPTITEEEFEDSLATDDFLVDYFNEFLSLPTFSEAIR...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>70</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13754</th>\n",
       "      <td>P13504</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>SP</td>\n",
       "      <td>MENMKVLLGLICLMVPLLSLEIDVCTEYPNQIVLFLSVNEIDIRKC...</td>\n",
       "      <td>SSSSSSSSSSSSSSSSSSSOOOOOOOOOOOOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13515</th>\n",
       "      <td>O18994</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>SP</td>\n",
       "      <td>MTVQLGAFYLFPLFMAGFVQTNSNLEKMDCYKDVTGTIYDYDAFTL...</td>\n",
       "      <td>SSSSSSSSSSSSSSSSSSSSSOOOOOOOOOOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15311</th>\n",
       "      <td>P80686</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>SP</td>\n",
       "      <td>MYKFVVFAAALAYANAGLIGAPAVAAYSAAPAVSSAYIHQAAPVAV...</td>\n",
       "      <td>SSSSSSSSSSSSSSSSOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14725</th>\n",
       "      <td>E3W9M2</td>\n",
       "      <td>EUKARYA</td>\n",
       "      <td>SP</td>\n",
       "      <td>MNMSCKFEIVLLVSWWLLLVLVFGVESSMFSEFDRLDFPKHFIFGA...</td>\n",
       "      <td>SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSOOOOOOOOOOOOOOOO...</td>\n",
       "      <td>70</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3419 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      uniprot_id   kingdom   type  \\\n",
       "18084     B2VGX7  NEGATIVE    TAT   \n",
       "119       Q9DEN4   EUKARYA  NO_SP   \n",
       "1049      A9QM74   EUKARYA  NO_SP   \n",
       "2541      Q8K3I4   EUKARYA  NO_SP   \n",
       "4434      Q920M2   EUKARYA  NO_SP   \n",
       "...          ...       ...    ...   \n",
       "4214      Q8NE09   EUKARYA  NO_SP   \n",
       "13754     P13504   EUKARYA     SP   \n",
       "13515     O18994   EUKARYA     SP   \n",
       "15311     P80686   EUKARYA     SP   \n",
       "14725     E3W9M2   EUKARYA     SP   \n",
       "\n",
       "                                                sequence  \\\n",
       "18084  MKAVNPLTENDVTPESLFNARRRTVLKMLGMSAAALSLPGAARADL...   \n",
       "119    MTLSGSGSASDMSGQTVLSADDADIDVVGEGDEALDKDSECESTAG...   \n",
       "1049   MPTLDAPEERRRKFKYRGKDVSLRRQQRMAVSLELRKAKKDEQTLK...   \n",
       "2541   MGRKLDLSGLTDDETEHVLQVVQRDFNLRKKEEDRLSEMKQRLAEE...   \n",
       "4434   MEVPELGPGLVERLEQLATCPLCGGPFEDPVLLACEHSFCRSCLAR...   \n",
       "...                                                  ...   \n",
       "4214   MPEKRLTAEPPTITEEEFEDSLATDDFLVDYFNEFLSLPTFSEAIR...   \n",
       "13754  MENMKVLLGLICLMVPLLSLEIDVCTEYPNQIVLFLSVNEIDIRKC...   \n",
       "13515  MTVQLGAFYLFPLFMAGFVQTNSNLEKMDCYKDVTGTIYDYDAFTL...   \n",
       "15311  MYKFVVFAAALAYANAGLIGAPAVAAYSAAPAVSSAYIHQAAPVAV...   \n",
       "14725  MNMSCKFEIVLLVSWWLLLVLVFGVESSMFSEFDRLDFPKHFIFGA...   \n",
       "\n",
       "                                                  labels  length  has_sp  \n",
       "18084  TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTOO...      70    True  \n",
       "119    IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "1049   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "2541   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "4434   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "...                                                  ...     ...     ...  \n",
       "4214   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...      70   False  \n",
       "13754  SSSSSSSSSSSSSSSSSSSOOOOOOOOOOOOOOOOOOOOOOOOOOO...      70    True  \n",
       "13515  SSSSSSSSSSSSSSSSSSSSSOOOOOOOOOOOOOOOOOOOOOOOOO...      70    True  \n",
       "15311  SSSSSSSSSSSSSSSSOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO...      70    True  \n",
       "14725  SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSOOOOOOOOOOOOOOOO...      70    True  \n",
       "\n",
       "[3419 rows x 7 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "e123f914304159f4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-17T21:23:44.701496Z",
     "start_time": "2025-10-17T21:23:44.633272Z"
    }
   },
   "outputs": [],
   "source": [
    "### save datasets\n",
    "reduced_df.to_csv(\"aufgabe3/reduced_30_signalP6_oversampled.csv\", index=False)\n",
    "train_df.to_csv(\"aufgabe3/reduced_30_signalP6_train_oversampled.csv\", index=False)\n",
    "val_df.to_csv(\"aufgabe3/reduced_30_signalP6_val_oversampled.csv\", index=False)\n",
    "test_df.to_csv(\"aufgabe3/reduced_30_signalP6_test_oversampled.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "o1mw94kg37n",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-16T19:03:27.838413Z",
     "start_time": "2025-10-16T19:03:27.827707Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created 3-Fold Cross Validation\n",
      "CV dataset size: 13,675 sequences\n",
      "Test set size: 3,419 sequences (held out)\n",
      "\n",
      "Fold 1:\n",
      "  Train: 9,116 samples (1,548 SP, 7,568 NO_SP)\n",
      "  Val:   4,559 samples (774 SP, 3,785 NO_SP)\n",
      "Fold 2:\n",
      "  Train: 9,117 samples (1,548 SP, 7,569 NO_SP)\n",
      "  Val:   4,558 samples (774 SP, 3,784 NO_SP)\n",
      "Fold 3:\n",
      "  Train: 9,117 samples (1,548 SP, 7,569 NO_SP)\n",
      "  Val:   4,558 samples (774 SP, 3,784 NO_SP)\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import StratifiedKFold\n",
    "\n",
    "# Setup 3-fold cross validation\n",
    "n_folds = 3\n",
    "skf = StratifiedKFold(n_splits=n_folds, shuffle=True, random_state=42)\n",
    "\n",
    "# Create fold indices\n",
    "fold_splits = list(skf.split(train_val_df, train_val_df['type']))\n",
    "\n",
    "print(f\"Created {n_folds}-Fold Cross Validation\")\n",
    "print(f\"CV dataset size: {len(train_val_df):,} sequences\")\n",
    "print(f\"Test set size: {len(test_df):,} sequences (held out)\\n\")\n",
    "\n",
    "# Show fold statistics\n",
    "for fold_idx, (train_idx, val_idx) in enumerate(fold_splits):\n",
    "    fold_train = train_val_df.iloc[train_idx]\n",
    "    fold_val = train_val_df.iloc[val_idx]\n",
    "    \n",
    "    print(f\"Fold {fold_idx + 1}:\")\n",
    "    print(f\"  Train: {len(fold_train):,} samples ({len(fold_train[fold_train['has_sp']]):,} SP, {len(fold_train[~fold_train['has_sp']]):,} NO_SP)\")\n",
    "    print(f\"  Val:   {len(fold_val):,} samples ({len(fold_val[fold_val['has_sp']]):,} SP, {len(fold_val[~fold_val['has_sp']]):,} NO_SP)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "uq5dzz8fd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-16T19:04:42.256570Z",
     "start_time": "2025-10-16T19:04:42.137543Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Fold 1 ---\n",
      "Class distribution before: {'NO_SP': np.int64(7568), 'SP': np.int64(1082), 'LIPO': np.int64(349), 'TAT': np.int64(90), 'PILIN': np.int64(18), 'TATLIPO': np.int64(9)}\n",
      "Target count (median): 219\n",
      "Class distribution after: {'NO_SP': np.int64(7568), 'SP': np.int64(1082), 'LIPO': np.int64(349), 'TATLIPO': np.int64(219), 'TAT': np.int64(219), 'PILIN': np.int64(219)}\n",
      "Result: 9,116 → 9,656 train, 4,559 val\n",
      "\n",
      "--- Fold 2 ---\n",
      "Class distribution before: {'NO_SP': np.int64(7569), 'SP': np.int64(1082), 'LIPO': np.int64(349), 'TAT': np.int64(89), 'PILIN': np.int64(18), 'TATLIPO': np.int64(10)}\n",
      "Target count (median): 219\n",
      "Class distribution after: {'NO_SP': np.int64(7569), 'SP': np.int64(1082), 'LIPO': np.int64(349), 'TATLIPO': np.int64(219), 'PILIN': np.int64(219), 'TAT': np.int64(219)}\n",
      "Result: 9,117 → 9,657 train, 4,558 val\n",
      "\n",
      "--- Fold 3 ---\n",
      "Class distribution before: {'NO_SP': np.int64(7569), 'SP': np.int64(1082), 'LIPO': np.int64(350), 'TAT': np.int64(89), 'PILIN': np.int64(18), 'TATLIPO': np.int64(9)}\n",
      "Target count (median): 219\n",
      "Class distribution after: {'NO_SP': np.int64(7569), 'SP': np.int64(1082), 'LIPO': np.int64(350), 'TATLIPO': np.int64(219), 'PILIN': np.int64(219), 'TAT': np.int64(219)}\n",
      "Result: 9,117 → 9,658 train, 4,558 val\n",
      "\n",
      "✓ All 3 folds saved!\n",
      "Test set saved separately: 3,419 samples\n"
     ]
    }
   ],
   "source": [
    "# Helper function for moderate oversampling (to median, not max)\n",
    "def oversample_to_median(df, target_col=\"type\", random_state=69):\n",
    "    \"\"\"Oversample minority classes to median class count (more moderate than max).\"\"\"\n",
    "    counts = df[target_col].value_counts()\n",
    "    target_count = int(counts.median())\n",
    "    \n",
    "    print(f\"Class distribution before: {dict(counts)}\")\n",
    "    print(f\"Target count (median): {target_count}\")\n",
    "    \n",
    "    dfs_to_concat = []\n",
    "    for label in counts.index:\n",
    "        subset = df[df[target_col] == label]\n",
    "        current_count = len(subset)\n",
    "        \n",
    "        if current_count < target_count:\n",
    "            # Oversample minority classes to reach median\n",
    "            oversampled = subset.sample(n=target_count, replace=True, random_state=random_state)\n",
    "            dfs_to_concat.append(oversampled)\n",
    "        else:\n",
    "            # Keep majority classes as-is (no undersampling)\n",
    "            dfs_to_concat.append(subset)\n",
    "    \n",
    "    result = pd.concat(dfs_to_concat, ignore_index=True)\n",
    "    result = result.sample(frac=1, random_state=42).reset_index(drop=True)\n",
    "    \n",
    "    print(f\"Class distribution after: {dict(result[target_col].value_counts())}\")\n",
    "    return result\n",
    "\n",
    "# Save each fold as separate CSV files with oversampled training data\n",
    "for fold_idx, (train_idx, val_idx) in enumerate(fold_splits):\n",
    "    # Get fold data\n",
    "    fold_train_df = train_val_df.iloc[train_idx].copy()\n",
    "    fold_val_df = train_val_df.iloc[val_idx].copy()\n",
    "    \n",
    "    # Apply moderate oversampling to training set only\n",
    "    print(f\"\\n--- Fold {fold_idx + 1} ---\")\n",
    "    fold_train_oversampled = oversample_to_median(fold_train_df)\n",
    "    \n",
    "    # Save oversampled train, original val (validation should not be oversampled)\n",
    "    fold_train_oversampled.to_csv(f\"aufgabe3/3-fold/fold_{fold_idx + 1}_train_oversampled.csv\", index=False)\n",
    "    fold_val_df.to_csv(f\"aufgabe3/3-fold/fold_{fold_idx + 1}_val_oversampled.csv\", index=False)\n",
    "    \n",
    "    print(f\"Result: {len(fold_train_df):,} → {len(fold_train_oversampled):,} train, {len(fold_val_df):,} val\")\n",
    "\n",
    "print(f\"\\n✓ All {n_folds} folds saved!\")\n",
    "print(f\"Test set saved separately: {len(test_df):,} samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "wcdkrsvc0rl",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-16T19:05:49.321168Z",
     "start_time": "2025-10-16T19:05:49.232096Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "======================================================================\n",
      "DATASET ORGANIZATION SUMMARY\n",
      "======================================================================\n",
      "\n",
      "📁 Complete reduced dataset: 17,094 sequences\n",
      "   └─ File: reduced_30_signalP6.csv\n",
      "\n",
      "📁 Test Set (20%, held out): 3,419 sequences\n",
      "   ├─ NO_SP: 2,838 (83.0%)\n",
      "   ├─ Has SP: 581 (17.0%)\n",
      "   └─ File: reduced_30_signalP6_test.csv\n",
      "\n",
      "📁 Cross-Validation Set (80%): 13,675 sequences\n",
      "   ├─ NO_SP: 11,353 (83.0%)\n",
      "   ├─ Has SP: 2,322 (17.0%)\n",
      "   └─ Split into 3 folds:\n",
      "\n",
      "      Fold 1:\n",
      "      ├─ Train: 9,656 samples (oversampled to median) → fold_1_train_oversampled.csv\n",
      "      └─ Val:   4,559 samples (original) → fold_1_val.csv\n",
      "      Fold 2:\n",
      "      ├─ Train: 9,657 samples (oversampled to median) → fold_2_train_oversampled.csv\n",
      "      └─ Val:   4,558 samples (original) → fold_2_val.csv\n",
      "      Fold 3:\n",
      "      ├─ Train: 9,658 samples (oversampled to median) → fold_3_train_oversampled.csv\n",
      "      └─ Val:   4,558 samples (original) → fold_3_val.csv\n",
      "\n",
      "======================================================================\n",
      "✓ All datasets created successfully!\n",
      "======================================================================\n"
     ]
    }
   ],
   "source": [
    "print(\"=\"*70)\n",
    "print(\"DATASET ORGANIZATION SUMMARY\")\n",
    "print(\"=\"*70)\n",
    "print(f\"\\n📁 Complete reduced dataset: {len(reduced_df):,} sequences\")\n",
    "print(f\"   └─ File: reduced_30_signalP6.csv\\n\")\n",
    "\n",
    "print(f\"📁 Test Set (20%, held out): {len(test_df):,} sequences\")\n",
    "print(f\"   ├─ NO_SP: {len(test_df[~test_df['has_sp']]):,} ({len(test_df[~test_df['has_sp']])/len(test_df)*100:.1f}%)\")\n",
    "print(f\"   ├─ Has SP: {len(test_df[test_df['has_sp']]):,} ({len(test_df[test_df['has_sp']])/len(test_df)*100:.1f}%)\")\n",
    "print(f\"   └─ File: reduced_30_signalP6_test.csv\\n\")\n",
    "\n",
    "print(f\"📁 Cross-Validation Set (80%): {len(train_val_df):,} sequences\")\n",
    "print(f\"   ├─ NO_SP: {len(train_val_df[~train_val_df['has_sp']]):,} ({len(train_val_df[~train_val_df['has_sp']])/len(train_val_df)*100:.1f}%)\")\n",
    "print(f\"   ├─ Has SP: {len(train_val_df[train_val_df['has_sp']]):,} ({len(train_val_df[train_val_df['has_sp']])/len(train_val_df)*100:.1f}%)\")\n",
    "print(f\"   └─ Split into {n_folds} folds:\\n\")\n",
    "\n",
    "for fold_idx in range(n_folds):\n",
    "    fold_train = pd.read_csv(f\"aufgabe3/3-fold/fold_{fold_idx + 1}_train_oversampled.csv\")\n",
    "    fold_val = pd.read_csv(f\"aufgabe3/3-fold/fold_{fold_idx + 1}_val.csv\")\n",
    "    print(f\"      Fold {fold_idx + 1}:\")\n",
    "    print(f\"      ├─ Train: {len(fold_train):,} samples (oversampled to median) → fold_{fold_idx + 1}_train_oversampled.csv\")\n",
    "    print(f\"      └─ Val:   {len(fold_val):,} samples (original) → fold_{fold_idx + 1}_val.csv\")\n",
    "print(\"\\n\" + \"=\"*70)\n",
    "print(\"✓ All datasets created successfully!\")\n",
    "print(\"=\"*70)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sp-prediction-nosync",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
